internal/ls/string_completions.go

```diff
@@ -0,0 +1,548 @@
+package ls
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/checker"
+	"github.com/microsoft/typescript-go/internal/compiler"
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/lsp/lsproto"
+	"github.com/microsoft/typescript-go/internal/printer"
+	"github.com/microsoft/typescript-go/internal/tspath"
+)
+
+type completionsFromTypes struct {
+	types           []*checker.StringLiteralType
+	isNewIdentifier bool
+}
+
+type completionsFromProperties struct {
+	symbols           []*ast.Symbol
+	hasIndexSignature bool
+}
+
+type completionsFromPaths = []*pathCompletion
+
+type pathCompletion struct {
+	name string
+	// ScriptElementKindScriptElement | ScriptElementKindDirectory | ScriptElementKindExternalModuleName
+	kind      ScriptElementKind
+	extension string
+	textRange *core.TextRange
+}
+
+// *completionsFromTypes | *completionsFromProperties | completionsFromPaths
+type stringLiteralCompletions = any
+
+func (l *LanguageService) getStringLiteralCompletions(
+	ctx context.Context,
+	file *ast.SourceFile,
+	position int,
+	contextToken *ast.Node,
+	compilerOptions *core.CompilerOptions,
+	program *compiler.Program,
+	preferences *UserPreferences,
+	clientOptions *lsproto.CompletionClientCapabilities,
+) *lsproto.CompletionList {
+	// !!! reference comment
+	if IsInString(file, position, contextToken) {
+		if contextToken == nil || !ast.IsStringOrNumericLiteralLike(contextToken) {
+			return nil
+		}
+		entries := l.getStringLiteralCompletionEntries(
+			ctx,
+			file,
+			contextToken,
+			position,
+			program,
+			preferences)
+		return l.convertStringLiteralCompletions(
+			ctx,
+			entries,
+			contextToken,
+			file,
+			position,
+			program,
+			compilerOptions,
+			preferences,
+			clientOptions,
+		)
+	}
+	return nil
+}
+
+func (l *LanguageService) convertStringLiteralCompletions(
+	ctx context.Context,
+	completion stringLiteralCompletions,
+	contextToken *ast.StringLiteralLike,
+	file *ast.SourceFile,
+	position int,
+	program *compiler.Program,
+	options *core.CompilerOptions,
+	preferences *UserPreferences,
+	clientOptions *lsproto.CompletionClientCapabilities,
+) *lsproto.CompletionList {
+	if completion == nil {
```

Comment:
This check is insufficient because `completion` might still be an interface-wrapped `nil`. I'll refactor to avoid this problem.